### PR TITLE
form: allow custom GetDefault method

### PR DESCRIPTION
**Pull request by rogpeppe**
This will make it feasible to add default that come from files
(and avoid printing the whole contents of the file too).

### CODE REVIEW

File: form/cmd/formtest/main.go [hunk 1]
```diff
@@ -40,7 +40,7 @@ The entered values will be displayed at the end.
 			"name": environschema.Attr{
 				Description: "Your full name.",
 				Type:        environschema.Tstring,
-				Mandatory: true,
+				Mandatory:   true,
```
Comment (mattyw) [inline form/cmd/formtest/main.go:43]: This implies gofmt wasn't run on save - which just makes me sick to my stomach

**Comment by mhilton** [comment]
:+1: LGTM

File: form/form.go [hunk 2]
```diff
@@ -99,6 +99,19 @@ type IOFiller struct {
 	// ShowDescriptions holds whether attribute descriptions
 	// should be printed as well as the attribute names.
 	ShowDescriptions bool
+
+	// GetDefault returns the default value for the given attribute,
+	// which must have been coerced using the given checker.
+	// If there is no default, it should return (nil, "", nil).
```
Comment (mattyw) [inline form/form.go:105]: it should return or does return?

File: form/form.go [hunk 3]
```diff
@@ -99,6 +99,19 @@ type IOFiller struct {
...
+	//
+	// The display return value holds the string to use
+	// to describe the value of the default. If it's empty,
+	// fmt.Sprint(val) will be used.
+	//
+	// If GetDefault returns an error, it will be printed as warning.
```
Comment (mattyw) [inline form/form.go]: as a warning

File: form/form_test.go [hunk 4]
```diff
@@ -583,6 +587,95 @@ var ioFillerTests = []struct {
...
+		},
+	},
+	filler: form.IOFiller{
+		GetDefault: func(attr form.NamedAttr, checker schema.Checker) (interface{}, string, error) {
+			return "hello", "", nil
+		},
+	},
+	expectIO: `
+	|Press return to select a default value, or enter - to omit an entry.
+	|a [hello]: »
```
Comment (mattyw) [inline form/form_test.go:607]: I'm starting to not like  »

Comment (rogpeppe) [inline form/form_test.go:607]: really? it seems quite nice to me. alternative suggestion?

Comment (urosj) [inline form/form_test.go:607]: +1, we don't need », just use :

**Comment by urosj** [comment]
In the forms, required fields usually have the \* at the end of the label, while the ones that don't have that can be skipped. Also, if a data in the field is required and user hasn't provided any, the field should just be repeated until the user enters the data or terminates the process.

**Comment by mattyw** [comment]
Fix my comment nitpicks then LGTM

File: form/form_test.go (hunk 4 above)
Comment (rogpeppe) [inline form/form_test.go:607]: : isn't good because it appears (many times) in the expected output.

File: form/cmd/formtest/main.go (hunk 1 above)
Comment (rogpeppe) [inline form/cmd/formtest/main.go:43]: some smelling salts dear?

File: form/form.go (hunk 2 above)
Comment (rogpeppe) [inline form/form.go:105]: I should "should" is right here - we're describing what you need to
do to make a valid GetDefault function. "does" would imply
that it always does that, which ain't necessarily so.

File: form/form.go (hunk 3 above)
Comment (rogpeppe) [inline form/form.go]: done

**Comment by rogpeppe** [comment]
We'll do the \* thing in a separate PR, if that's OK.

**Pull request merged**